modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -79,21 +79,333 @@ interface DomParser {
   getAttributeFilters: () => ParserFilter[];
   addNodeFilter: (name: string, callback: (nodes: AstNode[], name: string, args: ParserArgs) => void) => void;
   getNodeFilters: () => ParserFilter[];
-  filterNode: (node: AstNode) => AstNode;
   parse: (html: string, args?: ParserArgs) => AstNode;
 }
 
+// For internal parser use only: a summary of the nodes that have been parsed
+interface WalkResult {
+  readonly invalidChildren: AstNode[];
+  readonly matchedNodes: Record<string, AstNode[]>;
+  readonly matchedAttributes: Record<string, AstNode[]>;
+}
+
+const builder = (rootNode: AstNode, html: string, settings: DomParserSettings, schema: Schema, args: ParserArgs) => {
+  const blockElements = extend(makeMap('script,style,head,html,body,title,meta,param'), schema.getBlockElements());
+  const whiteSpaceElements = schema.getWhiteSpaceElements();
+  const nonEmptyElements = schema.getNonEmptyElements();
+  const allWhiteSpaceRegExp = /[ \t\r\n]+/g;
+  const isAllWhiteSpaceRegExp = /^[ \t\r\n]+$/;
+  const startWhiteSpaceRegExp = /^[ \t\r\n]+/;
+  const endWhiteSpaceRegExp = /[ \t\r\n]+$/;
+
+  let node = rootNode;
+  let isInWhiteSpacePreservedElement = Obj.has(whiteSpaceElements, args.context) || Obj.has(whiteSpaceElements, settings.root_name);
+
+  const removeWhitespaceBefore = (node: AstNode): void => {
+    const blockElements = schema.getBlockElements();
+
+    for (let textNode = node.prev; textNode && textNode.type === 3;) {
+      const textVal = textNode.value.replace(endWhiteSpaceRegExp, '');
+
+      // Found a text node with non whitespace then trim that and break
+      if (textVal.length > 0) {
+        textNode.value = textVal;
+        return;
+      }
+
+      const textNodeNext = textNode.next;
+
+      // Fix for bug #7543 where bogus nodes would produce empty
+      // text nodes and these would be removed if a nested list was before it
+      if (textNodeNext) {
+        if (textNodeNext.type === 3 && textNodeNext.value.length) {
+          textNode = textNode.prev;
+          continue;
+        }
+
+        if (!blockElements[textNodeNext.name] && textNodeNext.name !== 'script' && textNodeNext.name !== 'style') {
+          textNode = textNode.prev;
+          continue;
+        }
+      }
+
+      const sibling = textNode.prev;
+      textNode.remove();
+      textNode = sibling;
+    }
+  };
+
+  const cloneAndExcludeBlocks = (input: SchemaMap) => {
+    const output: SchemaMap = {};
+
+    for (const name in input) {
+      if (name !== 'li' && name !== 'p') {
+        output[name] = input[name];
+      }
+    }
+
+    return output;
+  };
+
+  const parser = SaxParser({
+    validate: settings.validate,
+    document: settings.document,
+    allow_html_data_urls: settings.allow_html_data_urls,
+    allow_svg_data_urls: settings.allow_svg_data_urls,
+    allow_script_urls: settings.allow_script_urls,
```

Comment:
These also do validation so this would need to be moved.

Reply:
There's also parts in `SaxParser` that handle sanitizing the attributes. I'm not sure if we can move them in this PR but it's something to keep in mind.

Reply:
This has been pushed to the next task